src/Compilers/Core/Portable/DiagnosticAnalyzer/CompilerPerTreeOptionsProvider.CompilerOptionSet.cs

```diff
@@ -0,0 +1,62 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using Microsoft.CodeAnalysis.Options;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.Diagnostics
+{
+    internal sealed partial class CompilerPerTreeOptionsProvider
+    {
+        internal class CompilerOptionSet : OptionSet, IEnumerable<KeyValuePair<Option<string>, string>>
+        {
+            private readonly ImmutableDictionary<Option<string>, string> _options;
+
+            private CompilerOptionSet(ImmutableDictionary<Option<string>, string> options)
+            {
+                _options = options;
+            }
+
+            public CompilerOptionSet()
+                : this(ImmutableDictionary.Create(
+                    CompilerOptionSetKeyComparer.Instance,
+                    StringOrdinalComparer.Instance))
+            { }
+
+            public override object GetOption(OptionKey optionKey)
+                => _options.TryGetValue((Option<string>)optionKey.Option, out string value)
```

Comment:
Another case where we are accepting a base type and forcing it to be a derived type on the implementation. This feels very wrong.